pytennis/tennis.py

```diff
@@ -0,0 +1,489 @@
+                        self.restart = True
+                        time.sleep(0.5)
+                        self.GeneralReward = False
+                        nextplayer = 'A'
+                    else:
+                        self.restart = False
+                        self.GeneralReward = True
+
+                    # Sample memories and use the target DQN to produce the target Q-Value
+                    X_state_val, X_action_val, rewards, X_next_state_val, continues = (
+                        self.AgentB.sample_memories(self.AgentB.batch_size))
+                    arr = [X_next_state_val]
+                    next_q_values = self.AgentB.model.predict(arr)
+                    max_next_q_values = np.max(
+                        next_q_values, axis=1, keepdims=True)
+                    y_val = rewards + continues * self.AgentB.discount_rate * max_next_q_values
+
+                    # Train the online DQN
+                    self.AgentB.model.fit(X_state_val, tf.keras.utils.to_categorical(
+                        X_next_state_val, num_classes=10), verbose=0)
+
+                    nextplayer = 'A'
+                    self.updateIter += 1
+                    # evaluate B
+
+                else:
+                    # Online DQN evaluates what to do
+                    q_valueB = self.AgentB.model.predict([stateB])
+                    actionB = self.AgentB.epsilon_greedy(q_valueB, iteration)
+
+                    # Online DQN plays
+                    obsB, rewardB, doneB, infoB = self.stepB(
+                        action=actionB, count=count)
+                    next_stateB = actionB
+
+                    # Let's memorize what just happened
+                    self.AgentB.replay_memory.append(
+                        (stateB, actionB, rewardB, next_stateB, 1.0 - doneB))
+                    tateB = next_stateB
+
+                if nextplayer == 'B':
+                    count += 1
+                else:
+                    count = 0
+
+            iteration += 1
+
+            # CHECK BALL MOVEMENT
+            self.DISPLAYSURF.blit(self.PLAYERA, (self.playerax, 50))
+            self.DISPLAYSURF.blit(self.PLAYERB, (self.playerbx, 650))
+            self.DISPLAYSURF.blit(self.ball, (self.ballx, self.bally))
+            self.DISPLAYSURF.blit(self.randNumLabelA, (20, 15))
+            self.DISPLAYSURF.blit(self.randNumLabelB, (450, 15))
+            # self.DISPLAYSURF.blit(self.randNumLabelIter, (50, 40))
+
+            # update last coordinate
+            # self.lastxcoordinate = self.ballx
+
+            pygame.display.update()
+            self.fpsClock.tick(self.FPS)
+
+            for event in pygame.event.get():
+
+                if event.type == QUIT:
+                    # self.AgentA.model.save('models/AgentA.h5')
+                    # self.AgentB.model.save('models/AgentB.h5')
+                    pygame.quit()
+                    sys.exit()
+
+
+# initialize the 2 agents.
+AgentA = DQN()
```

Comment:
please separate out agent part from the environment. Environments files should just expose functions, using which agent will interact with the environment.